lib/mapping/MapEditUtils.cpp

```diff
@@ -0,0 +1,371 @@
+ *
+ */
+#include "StdInc.h"
+
+#include "MapEditUtils.h"
+
+#include "../filesystem/Filesystem.h"
+#include "../JsonNode.h"
+#include "CMap.h"
+#include "CMapOperation.h"
+
+MapRect::MapRect() : x(0), y(0), z(0), width(0), height(0)
+{
+
+}
+
+MapRect::MapRect(int3 pos, si32 width, si32 height) : x(pos.x), y(pos.y), z(pos.z), width(width), height(height)
+{
+
+}
+
+MapRect MapRect::operator&(const MapRect& rect) const
+{
+	bool intersect = right() > rect.left() && rect.right() > left() &&
+		bottom() > rect.top() && rect.bottom() > top() &&
+		z == rect.z;
+	if (intersect)
+	{
+		MapRect ret;
+		ret.x = std::max(left(), rect.left());
+		ret.y = std::max(top(), rect.top());
+		ret.z = rect.z;
+		ret.width = std::min(right(), rect.right()) - ret.x;
+		ret.height = std::min(bottom(), rect.bottom()) - ret.y;
+		return ret;
+	}
+	else
+	{
+		return MapRect();
+	}
+}
+
+si32 MapRect::left() const
+{
+	return x;
+}
+
+si32 MapRect::right() const
+{
+	return x + width;
+}
+
+si32 MapRect::top() const
+{
+	return y;
+}
+
+si32 MapRect::bottom() const
+{
+	return y + height;
+}
+
+int3 MapRect::topLeft() const
+{
+	return int3(x, y, z);
+}
+
+int3 MapRect::topRight() const
+{
+	return int3(right(), y, z);
+}
+
+int3 MapRect::bottomLeft() const
+{
+	return int3(x, bottom(), z);
+}
+
+int3 MapRect::bottomRight() const
+{
+	return int3(right(), bottom(), z);
+}
+
+CTerrainSelection::CTerrainSelection(CMap* map) : CMapSelection(map)
+{
+
+}
+
+void CTerrainSelection::selectRange(const MapRect& rect)
+{
+	rect.forEach([this](const int3 pos)
+		{
+			this->select(pos);
+		});
+}
+
+void CTerrainSelection::deselectRange(const MapRect& rect)
+{
+	rect.forEach([this](const int3 pos)
+		{
+			this->deselect(pos);
+		});
+}
+
+void CTerrainSelection::setSelection(const std::vector<int3>& vec)
+{
+	for (auto pos : vec)
+		this->select(pos);
+}
+
+void CTerrainSelection::selectAll()
+{
+	selectRange(MapRect(int3(0, 0, 0), getMap()->width, getMap()->height));
+	selectRange(MapRect(int3(0, 0, 1), getMap()->width, getMap()->height));
+}
+
+void CTerrainSelection::clearSelection()
+{
+	deselectRange(MapRect(int3(0, 0, 0), getMap()->width, getMap()->height));
+	deselectRange(MapRect(int3(0, 0, 1), getMap()->width, getMap()->height));
+}
+
+CObjectSelection::CObjectSelection(CMap* map) : CMapSelection(map)
+{
+
+}
+
+const std::string TerrainViewPattern::FLIP_MODE_DIFF_IMAGES = "D";
+
+const std::string TerrainViewPattern::RULE_DIRT = "D";
+const std::string TerrainViewPattern::RULE_SAND = "S";
+const std::string TerrainViewPattern::RULE_TRANSITION = "T";
+const std::string TerrainViewPattern::RULE_NATIVE = "N";
+const std::string TerrainViewPattern::RULE_NATIVE_STRONG = "N!";
+const std::string TerrainViewPattern::RULE_ANY = "?";
+
+TerrainViewPattern::TerrainViewPattern() : diffImages(false), rotationTypesCount(0), minPoints(0)
+{
+	maxPoints = std::numeric_limits<int>::max();
+}
+
+TerrainViewPattern::WeightedRule::WeightedRule(std::string& Name) : points(0), name(Name)
```

Comment:
or maybe even rvalue would be better here

```suggestion
TerrainViewPattern::WeightedRule::WeightedRule(const std::string & Name) : points(0), name(Name)
```